File: QCTC/CTraX.f
C------------------------------------------------------------------------------
C    This code is part of the MondoSCF suite of programs for linear scaling
C    electronic structure theory and ab initio molecular dynamics.
C
C    Copyright (2004). The Regents of the University of California. This
C    material was produced under U.S. Government contract W-7405-ENG-36
C    for Los Alamos National Laboratory, which is operated by the University
C    of California for the U.S. Department of Energy. The U.S. Government has
C    rights to use, reproduce, and distribute this software.  NEITHER THE
C    GOVERNMENT NOR THE UNIVERSITY MAKES ANY WARRANTY, EXPRESS OR IMPLIED,
C    OR ASSUMES ANY LIABILITY FOR THE USE OF THIS SOFTWARE.
C
C    This program is free software; you can redistribute it and/or modify
C    it under the terms of the GNU General Public License as published by the
C    Free Software Foundation; either version 2 of the License, or (at your
C    option) any later version. Accordingly, this program is distributed in
C    the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
C    the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
C    PURPOSE. See the GNU General Public License at www.gnu.org for details.
C
C    While you may do as you like with this software, the GNU license requires
C    that you clearly mark derivative software.  In addition, you are encouraged
C    to return derivative works to the MondoSCF group for review, and possible
C    disemination in future releases.
C------------------------------------------------------------------------------
C    ECONOMIZED (FULLY FACTORED) CONTRACTION OF SP MULTIPOLE TENSORS
C    Author: Matt Challacombe
C==============================================================================
      SUBROUTINE CTraX77(LP,LQ,Cp,Sp,Cpq,Spq,Cq,Sq)
C
      REAL*8 Cp(0:*),Sp(0:*),Cpq(0:*),Spq(0:*),Cq(0:*),Sq(0:*)
      REAL*8 TMP,SGN,TMPC,TWO,TMPS,SGNL,SGNLM
C
      INTEGER L,LP,LQ,LL,M,LDX,K,KK,LLKK,LLKM,N,KDX,LKDX,ID
C
      ID(L)=L*(L+1)/2
C
C     ZERO
C
      IF(LP.EQ.0)THEN
         Tmp=0.0D0
         DO l=0,LQ
            ll=ID(l)
            Cp(0)=Cp(0)+Cpq(ll)*Cq(ll)+Spq(ll)*Sq(ll)            
            DO m=1,l
              ldx=ll+m
              Tmp=Tmp+Cpq(ldx)*Cq(ldx)+Spq(ldx)*Sq(ldx)
           ENDDO
        ENDDO
        Cp(0)=Cp(0)+2.0D0*Tmp
        RETURN
      ENDIF
C
C     ONE
C
      Sgn=1.0D0
      DO l=0,LP
         ll=ID(l)
         TmpC=0.0D0
         DO k=0,LQ
            kk=ID(k)
            llkk=ID(l+k)
            TmpC=TmpC+Cpq(llkk)*Cq(kk)
         ENDDO
         Cp(ll)=Cp(ll)+Sgn*TmpC
         Sgn=-Sgn
      ENDDO
C
C     TWO
C
      Two=-2.0D0
      DO l=1,LP
         ll=ID(l)
         DO k=0,LQ
            kk=ID(k)
            llkk=ID(l+k)
            DO m=1,l                   
               ldx=ll+m
               llkm=llkk+m
               Cp(ldx)=Cp(ldx)+Two*Cpq(llkm)*Cq(kk)
               Sp(ldx)=Sp(ldx)+Two*Spq(llkm)*Cq(kk)
            ENDDO
            
         ENDDO
         Two=-Two
      ENDDO 
C
C     THREE
C
      Two=2.0D0
      DO l=0,LP
         ll=ID(l)
         TmpC=0.0D0
         DO k=0,LQ
            kk=ID(k)
            llkk=ID(l+k)
            DO n=1,k
               kdx=kk+n
               lkdx=llkk+n
               TmpC=TmpC+Cpq(lkdx)*Cq(kdx)+Spq(lkdx)*Sq(kdx)
            ENDDO
         ENDDO
         Cp(ll)=Cp(ll)+Two*TmpC
         Two=-Two
      ENDDO 
C
C     FOUR
C
      Two=-2.0D0
      DO l=1,LP
         ll=ID(l)
         DO k=1,LQ
            kk=ID(k)
            llkk=ID(l+k)
            DO m=1,l                   
               ldx=ll+m
               llkm=llkk+m
               TmpC =0.0D0
               TmpS =0.0D0
               DO n=1,k
                  kdx=kk+n
                  lkdx=llkm+n
                  TmpC=TmpC+Spq(lkdx)*Sq(kdx)+Cpq(lkdx)*Cq(kdx)
                  TmpS=TmpS-Cpq(lkdx)*Sq(kdx)+Spq(lkdx)*Cq(kdx)
               ENDDO
               Cp(ldx)=Cp(ldx)+Two*TmpC
               Sp(ldx)=Sp(ldx)+Two*TmpS
            ENDDO
         ENDDO
         Two=-Two
      ENDDO 
C
C     FIVE
C
      SgnL=1.0D0
      DO l=1,LP
         ll=ID(l)
         DO k=1,LQ
            kk=ID(k)
            llkk=ID(l+k)
            SgnLM=SgnL
            DO m=1,l      
               ldx=ll+m
               llkm=llkk+m
               Two=2.0D0*SgnL   
               DO n=1,MIN(m,k)
                  kdx=kk+n
                  lkdx=llkm-n
                  Cp(ldx)=Cp(ldx)+Two*Cpq(lkdx)*Cq(kdx)
                  Sp(ldx)=Sp(ldx)+Two*Cpq(lkdx)*Sq(kdx)
                  Two=-Two
               ENDDO
               llkm=llkk-m
               TmpC=0.0D0
               TmpS=0.0D0
               DO n=m+1,k
                  kdx=kk+n
                  lkdx=llkm+n
                  TmpC=TmpC+Cpq(lkdx)*Cq(kdx)
                  TmpS=TmpS+Cpq(lkdx)*Sq(kdx)
               ENDDO
               Cp(ldx)=Cp(ldx)+SgnLM*2.0D0*TmpC
               Sp(ldx)=Sp(ldx)+SgnLM*2.0D0*TmpS
               SgnLM=-SgnLM
            ENDDO
         ENDDO
         SgnL=-SgnL
      ENDDO 
C
      SgnL=1.0D0
      DO l=1,LP
         ll=ID(l)
         DO k=1,LQ
            kk=ID(k)
            llkk=ID(l+k)
            SgnLM=SgnL
            DO m=1,l      
               ldx=ll+m
               llkm=llkk+m
               Two=-2.0D0*SgnL               
               DO n=1,MIN(m-1,k)
                  kdx=kk+n
                  lkdx=llkm-n            
                  Cp(ldx)=Cp(ldx)+Two*Spq(lkdx)*Sq(kdx)
                  Sp(ldx)=Sp(ldx)-Two*Spq(lkdx)*Cq(kdx)
                  Two=-Two
               ENDDO
               llkm=llkk-m
               TmpC =0.0D0
               TmpS =0.0D0
               DO n=m+1,k
                  kdx=kk+n
                  lkdx=llkm+n
                  TmpC=TmpC+Spq(lkdx)*Sq(kdx)
                  TmpS=TmpS+Spq(lkdx)*Cq(kdx)
               ENDDO
               Cp(ldx)=Cp(ldx)+SgnLM*2.0D0*TmpC
               Sp(ldx)=Sp(ldx)-SgnLM*2.0D0*TmpS          
               SgnLM=-SgnLM
            ENDDO
         ENDDO
         SgnL=-SgnL
      ENDDO 
C
      RETURN
      END

File: QCTC/CTraX/CTraX.m
(* 
!------------------------------------------------------------------------------
!--  This code is part of the MondoSCF suite of programs for linear scaling 
!    electronic structure theory and ab initio molecular dynamics.
!
!--  Copyright (c) 2001, the Regents of the University of California.  
!    This SOFTWARE has been authored by an employee or employees of the 
!    University of California, operator of the Los Alamos National Laboratory 
!    under Contract No. W-7405-ENG-36 with the U.S. Department of Energy.  
!    The U.S. Government has rights to use, reproduce, and distribute this 
!    SOFTWARE.  The public may copy, distribute, prepare derivative works 
!    and publicly display this SOFTWARE without charge, provided that this 
!    Notice and any statement of authorship are reproduced on all copies.  
!    Neither the Government nor the University makes any warranty, express 
!    or implied, or assumes any liability or responsibility for the use of 
!    this SOFTWARE.  If SOFTWARE is modified to produce derivative works, 
!    such modified SOFTWARE should be clearly marked, so as not to confuse 
!    it with the version available from LANL.  The return of derivative works
!    to the primary author for integration and general release is encouraged. 
!    The first publication realized with the use of MondoSCF shall be
!    considered a joint work.  Publication of the results will appear
!    under the joint authorship of the researchers nominated by their
!    respective institutions. In future publications of work performed
!    with MondoSCF, the use of the software shall be properly acknowledged,
!    e.g. in the form "These calculations have been performed using MondoSCF, 
!    a suite of programs for linear scaling electronic structure theory and
!    ab initio molecular dynamics", and given appropriate citation.  
!------------------------------------------------------------------------------
!    Author: Matt Challacombe
!    WRITE EXPLICIT CODE FOR MULTIPOLE GENERATION AND CONTRACTION
!------------------------------------------------------------------------------
*)

(* GET THE MAX ANGULAR SYMMETRY TO BE USED *)

MondoHome=Environment["FREEON_HOME"];
If[MondoHome==$FAILED,
   Print["COULD NOT FIND $FREEON_HOME! CHECK YOUR .cshrc "];
   Abort[];
  ];
EllFile = StringJoin[MondoHome,"/Includes/Ell.m"];
Get[EllFile];
(* Override of expansion length *)

(* INDEXING OF SP ARRAY ELEMENTS: NOTE WE START FROM 0 NOT 1*)

SPDex[L_,M_]:=Binomial[L+1,2]+M;
LSP[L_]:=L*(L+3)/2;   

(* LOAD OPTIMIZING AND FORMATING ROUTINES FOR MMA AND SET ASSOCIATED OPTIONS *)

Get[StringJoin[MondoHome,"/MMA/FixedNumberForm.m"]];
Get[StringJoin[MondoHome,"/MMA/Format.m"]];
Get[StringJoin[MondoHome,"/MMA/Optimize.m"]];

SetOptions[FortranAssign,AssignOptimize->False,AssignMaxSize->500,AssignPrecision -> Infinity,
                         AssignBreak->{132," & \n          "},AssignTemporary->{W,Array}]
SetOptions[Optimize,OptimizeVariable->{R,Sequence},OptimizeFunction->False]
SetOptions[OpenWrite, PageWidth -> 800];

(* GLOBAL FORMATING FUNCTIONS *)

FF[x_] := ToString[FixedNumberForm[SetPrecision[N[x,32],32], 16, 2]];
FF8[x_] := ToString[FixedNumberForm[SetPrecision[N[x,32],32], 6, 2]];
WS[String_]:=WriteString[FileName,"      ",String,"\n"];
WC[String_]:=WriteString[FileName,"!     ",String,"\n"];

(* LOAD MODULE FOR WRITING EXPLICIT SOURCE FOR COMPUTATION OF IRREGULAR HARMONICS *)

<<IrRegular.m; 
<<Contract.m;

Do[
   FileName=StringJoin["CTraX",ToString[LLSP],".F"]; 
   Print[" Openned ",FileName];
   OpenWrite[FileName];
   Do[
      LCode=100*LHG+LLSP;
      SubRoutineList=Contract[FileName,LCode,LHG,LLSP]; 
      IrRegular[FileName,LHG+LLSP,LCode];
      WS[StringJoin["SUBROUTINE CTraX",ToString[LCode],"(PQx,PQy,PQz,Cq,Sq,Cp,Sp)"]];
      WS[           "   IMPLICIT NONE"];
      WS[           "   REAL*8 PQx,PQy,PQz"];
      LDXP=ToString[LSP[LHG]];
      LDXQ=ToString[LSP[LLSP]];
      WS[StringJoin["   REAL*8 Cq(0:",LDXQ,"),Sq(0:",LDXQ,"),Cp(0:",LDXP,"),Sp(0:",LDXP,")"]];
      lensp=ToString[LSP[LHG+LLSP]];
      WS[StringJoin["   REAL*8 Cpq(0:",lensp,"),Spq(0:",lensp,")"]];
      WS[StringJoin["   CALL IrRegular",ToString[LCode],"(PQx,PQy,PQz,Cpq,Spq)"]];      
      Do[WS[SubRoutineList[[i]]];,{i,Length[SubRoutineList]}];
      WS[StringJoin["RETURN"]]
      WS[StringJoin["END"]]
   ,{LHG,0,Min[HGEll,LLSP-2]}];
   Close[FileName];
   Print[" Closed ",FileName];
,{LLSP,4,HGEll+14,2}]

(* NOW WRITE THE IRREGULARS FORTRAN77 FILE  *)

(* NOW WRITE THE F90 INCLUDES FILE FOR CALLS TO CTRAX *)

FileName="CTraX.Inc"; 
Print[" Openned ",FileName];
OpenWrite[FileName];
Print["hgell = ",HGEll];
Do[Do[
      ldx=ToString[LHG*100+LLSP];
      WS[StringJoin["CASE(",ldx,")"]];
      WS["   DO N=1,NFar"];
      WS["      Q=>Far(N)%P"];
      WS["      PQx=QC%Prim%P(1)-Q%Pole%Center(1)"];
      WS["      PQy=QC%Prim%P(2)-Q%Pole%Center(2)"];
      WS["      PQz=QC%Prim%P(3)-Q%Pole%Center(3)"];
      WS[StringJoin["      CALL CTraX",ldx,"(PQx,PQy,PQz,Q%Pole%C(0),Q%Pole%S(0),SPKetC(0),SPKetS(0))"]];
      WS["   ENDDO"];   
	 ,{LHG,0,Min[HGEll,LLSP-2]}];
      ,{LLSP,4,HGEll+14,2}];
   WS["CASE DEFAULT"];
WS["   STOP ' Unsuported expansion orders in TreeWalk.F90'"];
   WS["END SELECT"];
Close[FileName];
Print[" Closed ",FileName];

